File: sdk/ios/Sources/FrakSDK/Config/Backoff.swift
import Foundation

/// Per-key failure backoff: exponential, jittered, `Retry-After`-aware.
///
/// A backoff hit suppresses the network, not the answer — cached data is still served.
/// A plain struct, held as isolated state inside an owning actor rather than being
/// an actor itself.
struct Backoff {
    private struct Entry {
        let failureCount: Int
        let retryAt: Date
    }

    static let minDelay: TimeInterval = 1
    static let maxDelay: TimeInterval = 60
    private static let maxShift = 6

    private var state: [String: Entry] = [:]
    private let now: @Sendable () -> Date
    private let random: @Sendable (ClosedRange<Double>) -> Double

    init(
        now: @escaping @Sendable () -> Date = { Date() },
        random: @escaping @Sendable (ClosedRange<Double>) -> Double = { Double.random(in: $0) }
    ) {
        self.now = now
        self.random = random
    }

    /// True when `key` is inside its backoff window and must not be dialled.
    mutating func isBackingOff(_ key: String) -> Bool {
        guard let entry = state[key] else { return false }
        if now() >= entry.retryAt {
            // Expired windows are dropped on read, so the map cannot grow unbounded.
            state.removeValue(forKey: key)
            return false
        }
        return true
    }

    /// Records a failure and arms the next window. A server `Retry-After` — pulled out of
    /// `error` when it is a `.server` — acts as a floor, not a replacement: the exponential
    /// must still grow past repeated failures.
    mutating func recordFailure(_ key: String, from error: FrakError?) {
        let failureCount = (state[key]?.failureCount ?? 0) + 1
        let exponential = Self.minDelay * pow(2, Double(min(failureCount - 1, Self.maxShift)))
        let capped = min(exponential, Self.maxDelay)
        let retryAfterSeconds: Int?
        if case .server(_, _, let seconds) = error {
            retryAfterSeconds = seconds
        } else {
            retryAfterSeconds = nil
        }
        let serverFloor = TimeInterval(retryAfterSeconds ?? 0)
        let delay = max(capped, serverFloor)
        state[key] = Entry(failureCount: failureCount, retryAt: now().addingTimeInterval(jitter(delay)))
    }

    mutating func recordSuccess(_ key: String) {
        state.removeValue(forKey: key)
    }

    /// Records `error` against `key` and rethrows it. For a non-2xx already mapped to a
    /// `FrakError`.
    mutating func recordFailureAndThrow(_ key: String, _ error: FrakError) throws -> Never {
        recordFailure(key, from: error)
        throw error
    }

    /// Runs `request`, recording a failure and rethrowing on `FrakError`.
    ///
    /// A free function rather than an instance method: Swift does not allow an `async`
    /// mutating method to be called on an actor-isolated stored property, since the
    /// exclusive access it needs would have to span the `await`. `onFailure` is invoked
    /// synchronously in the `catch`, so it is safe for a caller to mutate its own
    /// actor-isolated `Backoff` from there. Inherits the caller's isolation via
    /// `#isolation`, so passing it the caller's own actor-isolated closures needs no hop.
    static func runOrRecordFailure<T>(
        isolation: isolated (any Actor)? = #isolation,
        _ request: () async throws -> T,
        onFailure: (FrakError) -> Void
    ) async throws -> T {
        do {
            return try await request()
        } catch let error as FrakError {
            onFailure(error)
            throw error
        }
    }

    private func jitter(_ delay: TimeInterval) -> TimeInterval {
        guard delay > 0 else { return 0 }
        let half = delay / 2
        return random(half...delay)
    }
}

File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/config/Backoff.kt
package id.frak.sdk.config

import id.frak.sdk.core.FrakError
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.math.min
import kotlin.random.Random

/**
 * Per-key failure backoff: exponential, jittered (full range `[delay/2, delay]`),
 * `Retry-After`-aware. Suppresses the network, not the answer; cached data still served. Not
 * thread-safe by itself; callers hold it behind their own mutex.
 */
internal class Backoff(
    private val now: () -> Long = System::currentTimeMillis,
    private val random: Random = Random.Default,
) {
    private val state = HashMap<String, Entry>()

    private data class Entry(
        val failureCount: Int,
        val retryAtMillis: Long,
    )

    /** True when [key] is inside its backoff window and must not be dialled. */
    fun isBackingOff(key: String): Boolean = remainingMillis(key) != null

    /**
     * What is left of [key]'s backoff window, or null when it isn't backing off. Callers surface
     * this as [FrakError.BackingOff.retryAfterMillis]; it is always positive.
     */
    fun remainingMillis(key: String): Long? {
        val entry = state[key] ?: return null
        val remaining = entry.retryAtMillis - now()
        if (remaining <= 0L) {
            // Dropped on read rather than swept, so the map can't grow unbounded over a session.
            state.remove(key)
            return null
        }
        return remaining
    }

    /** `Retry-After` is a floor, not a replacement: the exponential must still grow past it. */
    fun recordFailure(
        key: String,
        error: Throwable?,
    ) {
        val failureCount = (state[key]?.failureCount ?: 0) + 1
        val exponential = MIN_DELAY_MILLIS shl min(failureCount - 1, MAX_SHIFT)
        val capped = min(exponential, MAX_DELAY_MILLIS)
        val serverFloor = (error as? FrakError.Server)?.retryAfterSeconds?.times(1_000L) ?: 0L
        val delay = maxOf(capped, serverFloor)
        state[key] = Entry(failureCount, now() + jitter(delay))
    }

    /** Clears the window for [key]. Called on every success. */
    fun recordSuccess(key: String) {
        state.remove(key)
    }

    /**
     * Runs [request], recording a failure against [key] under [mutex] and rethrowing. Callers
     * still record their own success, since that happens alongside writing their own cache entry.
     */
    suspend fun <T> runOrRecordFailure(
        mutex: Mutex,
        key: String,
        request: suspend () -> T,
    ): T =
        try {
            request()
        } catch (failure: FrakError) {
            mutex.withLock { recordFailure(key, failure) }
            throw failure
        }

    /** Records [error] against [key] under [mutex] and throws it. For a non-2xx already mapped to a [FrakError]. */
    suspend fun recordFailureAndThrow(
        mutex: Mutex,
        key: String,
        error: FrakError,
    ): Nothing {
        mutex.withLock { recordFailure(key, error) }
        throw error
    }

    private fun jitter(delayMillis: Long): Long {
        if (delayMillis <= 0L) return 0L
        val half = delayMillis / 2
        return half + random.nextLong(half + 1)
    }

    companion object {
        const val MIN_DELAY_MILLIS: Long = 1_000
        const val MAX_DELAY_MILLIS: Long = 60_000

        /** Caps the shift so `1s shl n` cannot overflow before the cap bites. */
        private const val MAX_SHIFT = 6
    }
}
